ai_diffusion/ui/widget.py

```diff
@@ -590,7 +591,66 @@ def move_cursor_to_end(self):
             self._single.setCursorPosition(len(self._single.text()))
 
 
+class StrengthSnapping:
+    model: Model
+
+    def __init__(self, model: Model):
+        self.model = model
+
+    def get_steps(self) -> tuple[int, int]:
+        if self.model.workspace is Workspace.generation:
+            is_live = False
+        elif self.model.workspace is Workspace.upscaling:
+            assert False
+        elif self.model.workspace is Workspace.live:
+            is_live = True
+        elif self.model.workspace is Workspace.animation:
+            is_live = self.model.animation.sampling_quality is SamplingQuality.fast
+        else:
+            assert False, "unknown type of workspace"
+        return self.model.style.get_steps(is_live=is_live)
+
+    def nearest_percent(self, value: int) -> int | None:
+        min_steps, max_steps = self.get_steps()
+        strength = value / 100
+        steps, start_at_step = apply_strength(strength, steps=max_steps, min_steps=min_steps)
+        return snap_to_percent(steps, start_at_step, max_steps=max_steps)
+
+    def apply_strength(self, value: int) -> tuple[int, int]:
```

Comment:
looks like `nearest_percent` should make use of this function

Reply:
Not totally clean cause we still need `max_steps`. But done.